### ISSUE TITLE
Use reference library without corresponding runtime implementation

### ISSUE THREAD

**Issue by andschwa** [dotnet/cli#1043]
For an internal Microsoft project, we're building a legacy codebase across both Windows and Linux (and OS X) using `dotnet-cli`. Unfortunately, with the changes to `dotnet-cli` to use NuGet instead of DNX, we're crashing into some "guardrails" so to speak.

To this point, to port the codebase, we've added runtime guards around code that can't (and will never) work on Linux, but we still require the types (via reference assemblies) in order to compile. CLI using NuGet has added compatibility checks that DNX previously didn't care about, which is now preventing us from restoring necessary packages.

For now, we can lock to .NET packages at 1.0.0-rc2-23712, which appear to not have those checks enforced. Unfortunately, this is less than ideal because we a) would really like to to pull in new packages to test fixes for some issues, and b) the more behind we get, the harder it gets to upgrade.

So our project has taken a dependency on a feature that was quite officially supported; that is, the ability to use a reference assembly and intentionally not caring about its lack of implementation (on particular platforms).

@anurse has suggested a `--no-guardrails` option to give us this feature in new versions of `dotnet-cli` with new versions of the .NET packages.

/cc @ericstj @vors @FawcettJohnW

**Comment by ericstj** [comment]
The difference with the 1.0.0-rc2-23712 packages is that System.Security.Principal.Windows is now packaged as a Windows-specific implementation.  https://github.com/dotnet/corefx/issues/4925  We do intend to fix up more of the packages System.*.AccessControl, MS.Win32.Registry, etc to do the same.  /cc @chcosta

**Comment by anurse** [comment]
To be clear, the `--no-guardrails` suggestion is purely a tactical fix. I don't think it's a good solution long term, since as @ericstj has pointed out before, it's good to have guardrails around. The suggestion was just to allow advanced usages to work around packaging issues.

**Comment by andschwa** [comment]
The other packages @ericstj we also have to use (for the types on Linux, and implementation on Windows).

I am not sure what the best solution is going to be for this, but I'm sure we'll come up with a way of doing it right. I do *want* guardrails, I totally understand having them, but yes, porting requires strange things.

**Comment by piotrpMSFT** [comment]
I think the plan here is to:
1. Use --no-guardrails to get unblocked in the short term
2. Fix up NETStandard.Library dependencies a la  System.Security.Principal.Windows
3. Remove --no-guardrails

Do I have that right? If so, it seems we can link to a CoreFx repo issue and close this one out. Otherwise, what work do we need to do here?

@ericstj assigning to you, assuming my interpretation is right.

**Comment by andschwa** [comment]
@piotrpMSFT Just to be clear, it is not `NETStandard.Library` that's causing me an issue. I really do need to pull in packages without runtime implementations; it's that or rewrite a lot of code that I don't own. So we'd continue to need `--no-guardrails` until there exists a better way to do the same thing.

**Comment by davidfowl** [comment]
I don't think we should add a command line flag for this. I agree we need to do something but there are other options.

**Comment by andschwa** [comment]
I agree. I'd rather use something like:

```json
"frameworks": { "dnxcore50": { "allowRefOnly": true } }
```

For porting projects like ours where we just need to satisfy types at compile time.

**Comment by andschwa** [comment]
I have no clue who fixed it, but latest `dotnet-nightly` *and* latest 23729 build of .NET Core is working just fine in our scenario now. Thank you very much :smile:

**Comment by andschwa** [comment]
Okay, just FYI. With `dnxcore50`, and 23729, we can restore the packages. When using `netcore50`, we cannot, so we can't switch yet.

**Comment by andschwa** [comment]
I'm not even convinced `netcore50` is what we're supposed to eventually switch to; I think it's `netstandard1.4` but documentation is lacking. Tracked and will revisit later when I have more information.

**Comment by ericstj** [comment]
Docs: https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-standard-applications.md

You can switch to netstandardapp1.5 today but you'll need to have an import for dnxcore50 until we finish updating all the packages.

**Comment by andschwa** [comment]
These are the docs I have been needing @ericstj, thank you! I will switch tomorrow morning. I'm excited! :volcano:

**Comment by GuardRex** [comment]
@ericstj 

> You can switch to netstandardapp1.5 today

On a lark, I tried ...
```json
    "frameworks": {
        "nestandard1.5": {
            "imports": "portable-net45+win8"
        }
    },
```
... and ...
```json
    "frameworks": {
        "nestandard1.5": {
            "imports": "dnxcore50"
        }
    },
```
... and the output was similar on both tests ... many packages :smile: and many packages :frowning:

![capture1](https://cloud.githubusercontent.com/assets/1622880/12985709/d4773e10-d0a8-11e5-81f3-f6bcba9a2778.PNG)

**Comment by anurse** [comment]
Some packages (particularly the ones I see in your list) also use the old `dotnet` name :). It's a bit of whack-a-mole, but the following might just work:

```json
    "frameworks": {
        "netstandard1.5": {
            "imports": [ "dotnet5.6", "dnxcore50", "portable-net45+win8" ]
        }
    },
```

(support for multiple items in `imports` and non-portable profiles in `imports` is only in some of the most recent NuGet builds, it may not be there in the latest dotnet-cli builds)

**Comment by andschwa** [comment]
@anurse I like it, that's the kind of solution I'm used to lately :smiling_imp:

**Comment by GuardRex** [comment]
@anurse Similar result (about 50 packages that it choked on), but that was interesting just to see what it would do.

**Comment by ericstj** [comment]
I suspect you don't have imports support in your restore tool.  Try the latest nuget.exe and see if you get different results.  You can get that out of the nuget.commandline package, eg: https://www.myget.org/F/nugetbuild/api/v2/package/NuGet.CommandLine/3.4.0-beta-619.

FWIW dotnet5.6 and dnxcore50 are redundant.

Here is what I'd expect folks to do for compatibility.

In a library previously building as dotnet or dotnetX.Y:
```
    "frameworks": {
        "netstandard1.5": {
            "imports": [ "dotnet5.6" ]
        }
    },
```

In an application previously targeting DNXCore50
```
    "frameworks": {
        "netstandardapp1.5": {
            "imports": [ "dnxcore50", "portable-net45+win8" ]
        }
    },
```

@anurse if this is missing from the CLI we need to get that fixed ASAP.

**Comment by anurse** [comment]
@ericstj It's in the queue https://github.com/dotnet/cli/pull/1332 but we've been blocked getting new builds out for a while because of some CoreCLR bugs that are fixed. We're integrating 23811 now and that should open the floodgates for a few of these changes that have been blocked :)

**Comment by ericstj** [comment]
Cool, thanks @anurse

**Comment by GuardRex** [comment]
#### Simply re-restoring about three times back-to-back (tooling: `-001290-beta`) resulted in ...

![capture1](https://cloud.githubusercontent.com/assets/1622880/12988862/9bb936b4-d0b7-11e5-82c0-bfb8175f48b5.PNG)

#### ... and it also worked with your suggestion ...

![capture2](https://cloud.githubusercontent.com/assets/1622880/12989306/bad7903e-d0b9-11e5-882b-c5e3ee6636f7.PNG)

#### ... I haven't tried to run it yet, but I will shortly.

**Comment by GuardRex** [comment]
:smile: ... restore :+1: ... compile/publish :+1: ... runs under IIS :+1:. Tooling is `1.0.0.001290 beta`
```json
"dependencies": {
    "Microsoft.AspNetCore.Antiforgery": "1.0.0-*",
    "Microsoft.AspNetCore.HttpOverrides": "1.0.0-*",
    "Microsoft.AspNetCore.Diagnostics": "1.0.0-*",
    "Microsoft.AspNetCore.IISPlatformHandler": "1.0.0-*",
    "Microsoft.AspNetCore.Mvc": "1.0.0-*",
    "Microsoft.AspNetCore.Mvc.Razor": "1.0.0-*",
    "Microsoft.AspNetCore.Mvc.TagHelpers": "1.0.0-*",
    "Microsoft.AspNetCore.Server.Kestrel": "1.0.0-*",
    "Microsoft.AspNetCore.StaticFiles": "1.0.0-*",
    "Microsoft.AspNetCore.Tooling.Razor": "1.0.0-*",
    "Microsoft.Extensions.Configuration": "1.0.0-*",
    "System.Runtime.Serialization.Primitives": "4.1.0-*",
    "NETStandard.Library": "1.0.0-*",
    "System.Net.Http": "4.0.1-*",
    "System.Net.Security": "4.0.0-*",
    "System.Net.Sockets": "4.1.0-*",
    "WindowsAzure.Storage": "6.2.2-preview",
    "Microsoft.Extensions.Logging": "1.0.0-*",
    "Microsoft.Extensions.Logging.Console": "1.0.0-*",
    "Microsoft.Extensions.Logging.Tracesource": "1.0.0-*",
    "Microsoft.Extensions.DependencyModel": "1.0.0-dev-001200"
},
"frameworks": {
    "netstandardapp1.5": {
        "imports": [ "dnxcore50", "portable-net45+win8" ]
    }
},
```
... so when the time comes, I should be able to drop the `imports`.

I've had trouble getting `Microsoft.Extensions.DependencyModel 1.0.0-dev-001200` to float with `-dev-*`; but perhaps, that will clear up as well (or get dropped ... I'm not clear on if that's a temporary workaround package or something we'll be using permanently).

**Comment by GuardRex** [comment]
One not-so-good side effect though ... Intellisense is apparently needing some `dnx451` bits or something that aren't getting restored, so Intellisense is giving me the finger ... and not the good finger (:+1:).

![capture](https://cloud.githubusercontent.com/assets/1622880/12992988/7626995e-d0cc-11e5-87e3-5a3312631c60.PNG)

**Comment by andschwa** [comment]
Haha... so I have it working on OS X. Oddly. But my CI for Windows and Linux is failing; my guess is that their dotnet nightlies don't yet support `"imports": [ "dnxcore50", "portable-net45+win8" ]`.

I'll wait for new ones to drop :droplet:

**Comment by GuardRex** [comment]
It was working for me early in the day ... started failing mid-day ... I can't restore without hitting corrupt .dat files and 'end of directory tree' errors. Prob due to package churn. I'll try again tomorrow.

**Comment by andschwa** [comment]
This issue has resurfaced today. WIth the latest builds of CLI, our CI has started failing because it refuses to restore packages that don't support our runtime, so this is now blocking us (again). Any work-around?

/cc @brthor @piotrpMSFT

**Comment by piotrpMSFT** [comment]
Working with @andschwa 

@blackdwarf @krwq the issue we're seeing from @andschwa is partially due to our installation guidance. Specifically, we need to add a section that tells folks how to use CLI from CI.

**Comment by andschwa** [comment]
With .NET CLI 1.0.0 and some refactoring of our `project.json` files, this is now totally supported (please feel free to close the issue).

This is the structure we're using, tested on Linux, OS X, and Ubuntu for a project that was originally ported from FullCLR (and still runs there too, through a lot of magic).

## Top-level dependency that creates an executable

The key to be able to restore and build against all the dependencies (even those unsupported on Linux) is to use the new `netstandardapp1.5` moniker and transform this into a [.NET Standard Application](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-standard-applications.md) (with `"imports"` to support packages targetting older but compatible frameworks).

Additionally, a `"runtimes"` section *must* be added with RIDs for each platform you support.
 
```json
{
    "name": "Cross.Platform.App",
    "version": "1.0.0-*",
    "description": "Compile against Windows-only assemblies on Linux",

    "compilationOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "Cross.Platform.Library": "1.0.0-*"
    },

    "frameworks": {
        "netstandardapp1.5": {
            "imports": [ "dnxcore50", "portable-net45+win8" ]
        }
    },

    "runtimes": {
        "ubuntu.14.04-x64": { },
        "win7-x64": { },
        "win10-x64": { },
        "osx.10.10-x64": { },
        "osx.10.11-x64": { }
    }
}
```
## Some library that is also built

Correct me if I'm wrong, but the framework lacks "app" here because it does not get directly published; instead its target runtimes and frameworks are transitively determined by its own dependencies (in this case, NETStandard.Library).

We also had to change how we were specifying "sister" project dependencies (other libraries we build at the same time). I think we'd originally done it by example, so we had a `{ "type": "build" }` instead of just a package version number (see above).

```json
{
    "name": "Cross.Platform.Library",
    "version": "1.0.0-*",
    "description": "Some build dependency of the App",

    "dependencies": {
        "NETStandard.Library": "1.0.0-rc3-23818",
    },

    "frameworks": {
        "netstandard1.5": {
            "imports": [ "dnxcore50", "portable-net45+win8" ]
        }
    }
}
```

Thanks for all your help yesterday @piotrpMSFT and @emgarten! CLI is awesome :volcano:

**Comment by blackdwarf** [comment]
@piotrpMSFT not a problem, we are working on it!

**Comment by andschwa** [comment]
Unfortunately, I have to reopen this. I think it's more due to .NET Core updating their monikers to `netstandard1.5`, and NuGet not dealing with package that are required on a per-platform basis.

After bumping the package versions, I am greeted with an unfortunate set of errors:

```
Errors in C:\Users\andschwa\src\my-cross-platform-project\project.json
    Microsoft.Win32.Registry.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for Microsoft.Win32.Registry.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.10-x64.
    Microsoft.Win32.Registry 4.0.0-rc3-23902 provides a compile-time reference assembly for Microsoft.Win32.Registry on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.10-x64.
    System.ServiceProcess.ServiceController 4.1.0-rc3-23902 provides a compile-time reference assembly for System.ServiceProcess.ServiceController on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.10-x64.
    System.IO.FileSystem.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.10-x64.
    System.Security.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for System.Security.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.10-x64.
    Some packages are not compatible with .NETStandardApp,Version=v1.5 (osx.10.10-x64).
    Microsoft.Win32.Registry.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for Microsoft.Win32.Registry.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.11-x64.
    Microsoft.Win32.Registry 4.0.0-rc3-23902 provides a compile-time reference assembly for Microsoft.Win32.Registry on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.11-x64.
    System.ServiceProcess.ServiceController 4.1.0-rc3-23902 provides a compile-time reference assembly for System.ServiceProcess.ServiceController on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.11-x64.
    System.IO.FileSystem.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.11-x64.
    System.Security.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for System.Security.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with osx.10.11-x64.
    Some packages are not compatible with .NETStandardApp,Version=v1.5 (osx.10.11-x64).
    Microsoft.Win32.Registry.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for Microsoft.Win32.Registry.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with ubuntu.14.04-x64.
    Microsoft.Win32.Registry 4.0.0-rc3-23902 provides a compile-time reference assembly for Microsoft.Win32.Registry on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with ubuntu.14.04-x64.
    System.ServiceProcess.ServiceController 4.1.0-rc3-23902 provides a compile-time reference assembly for System.ServiceProcess.ServiceController on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with ubuntu.14.04-x64.
    System.IO.FileSystem.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with ubuntu.14.04-x64.
    System.Security.AccessControl 4.0.0-rc3-23902 provides a compile-time reference assembly for System.Security.AccessControl on .NETStandardApp,Version=v1.5, but there is no run-time assembly compatible with ubuntu.14.04-x64.
    Some packages are not compatible with .NETStandardApp,Version=v1.5 (ubuntu.14.04-x64).
```

**Comment by andschwa** [comment]
That's on:
```
Product Information:
 Version:     1.0.0-beta-001598
 Commit Sha:  7582649f88

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 Runtime Id:  win10-x64
```

And similarly on:

```
Product Information:
 Version:     1.0.0-beta-001610
 Commit Sha:  N/A

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  14.04
 OS Platform: Linux
 Runtime Id:  ubuntu.14.04-x64
```

**Comment by andschwa** [comment]
This is going to be (with pretty much the next release) blocking me from integrating much needed changes from .NET Core.

**Comment by GuardRex** [comment]
@andschwa Some troubleshooting advice has included the suggestion to use `netstandardapp1.5` ... not `netstandard1.5`. What that something you tried?

**Comment by emgarten** [comment]
Using nuget's Include/exclude for these packages seems like the right way to fix this. Currently those do not change the compatibility check result, but changing that may be the answer here.

**Comment by ericstj** [comment]
Yeah, I have a change prepped to use that in all of the corefx packages.  If those are the ones bringing in the refs then that'll fix it.

I don't think that's the case here since most of these are leaf, but a similar fix in your project might work.  In the failing project add "exclude": "compile" to the references that can't be resolved on OSX.

**Comment by andschwa** [comment]
@emgarten Yes, if it changes the compatibility check result, that would work.

@ericstj Yes, these are mostly leaf packages. However, I'm not sure what you mean. I do need them to compile and include the runtime implementation when I'm building on platforms that have support (Windows runtimes), and then in the same project, on platforms where there is no corresponding runtime (OS X, Ubuntu, and soon, CentOS/RHEL), I need the reference libraries restored so that the project compiles (because the types are still used), but do not need the runtime exceptions (because when those code paths are called, we throw a `PlatformNotSupportedException`). So is `"exclude": "compile"` something I could add to each runtime for each unsupported dependency...?

**Comment by emgarten** [comment]
I have a fix in NuGet for this in https://github.com/NuGet/Home/issues/2242

Adding ``"include": "compile"`` will allow NuGet to reference *only* the compile assets and skip the check for runtime assemblies where this is failing.

**Comment by ericstj** [comment]
I'm not sure if that will fix it.  Won't that prevent the implementation from being deployed on Windows as well?

**Comment by davidfowl** [comment]
It won't fix it.

> Won't that prevent the implementation from being deployed on Windows as well?

Yes

**Comment by andschwa** [comment]
That's what I was afraid of [here](https://github.com/NuGet/Home/issues/2242#issuecomment-191936562). :boom:

**Comment by andschwa** [comment]
I think this discussion should move to CoreFX, as it is more of an architectural issue than a build issue. CLI is operating as expected, and what's expected doesn't quite work for us.

Thanks everyone for your help thus far!

**Comment by andschwa** [comment]
CoreFX has taken care of this.
